reco_utils/recommender/lightgbm/lightgbm_utils.py

```diff
@@ -0,0 +1,224 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+import os, logging
+from sklearn.metrics import (
+    roc_auc_score,
+    log_loss,
+    mean_squared_error,
+    accuracy_score,
+    f1_score,
+)
+import numpy as np
+import pandas as pd
+import category_encoders as ce
+from tqdm import tqdm
+import collections
+import gc
+
+import tarfile
+from reco_utils.dataset.url_utils import maybe_download
+
+logging.basicConfig(level = logging.INFO, format = '%(asctime)s [INFO] %(message)s')
+
+def download_lgb_resources(azure_container_url, data_path, remote_resource_name):
+    """Download resources.
+
+    Args:
+        azure_container_url (str): URL of Azure container.
+        data_path (str): Path to download the resources.
+        remote_resource_name (str): Name of the resource.
+    """
+    os.makedirs(data_path, exist_ok=True)
+    remote_path = azure_container_url + remote_resource_name
+    maybe_download(remote_path, remote_resource_name, data_path)
+    tar_ref = tarfile.open(os.path.join(data_path, remote_resource_name), "r")
+    tar_ref.extractall(data_path)
+    tar_ref.close()
+    os.remove(os.path.join(data_path, remote_resource_name))
+
+def cal_metric(labels, preds, metrics):
+    """Calculate metrics,such as auc, logloss
+    FIXME: refactor this with the reco metrics
+    """
+    res = {}
+    for metric in metrics:
+        if metric == "auc":
+            auc = roc_auc_score(np.asarray(labels), np.asarray(preds))
+            res["auc"] = round(auc, 4)
+        elif metric == "rmse":
+            rmse = mean_squared_error(np.asarray(labels), np.asarray(preds))
+            res["rmse"] = np.sqrt(round(rmse, 4))
+        elif metric == "logloss":
+            # avoid logloss nan
+            preds = [max(min(p, 1.0 - 10e-12), 10e-12) for p in preds]
+            logloss = log_loss(np.asarray(labels), np.asarray(preds))
+            res["logloss"] = round(logloss, 4)
+        elif metric == "acc":
+            pred = np.asarray(preds)
+            pred[pred >= 0.5] = 1
+            pred[pred < 0.5] = 0
+            acc = accuracy_score(np.asarray(labels), pred)
+            res["acc"] = round(acc, 4)
+        elif metric == "f1":
+            pred = np.asarray(preds)
+            pred[pred >= 0.5] = 1
+            pred[pred < 0.5] = 0
+            f1 = f1_score(np.asarray(labels), pred)
+            res["f1"] = round(f1, 4)
+        else:
+            raise ValueError("not define this metric {0}".format(metric))
+    return res
+
+def unpackbits(x,num_bits):
+    xshape = list(x.shape)
+    x = x.reshape([-1,1])
+    to_and = 2**np.arange(num_bits).reshape([1,num_bits])
+    return (x & to_and).astype(bool).astype(int).reshape(xshape + [num_bits])
+
+class NumEncoder(object):
```

Comment:
can you provide some detail on what this class is doing and how it is used. I like that you're leveraging the categorical_encoder package, it's really nice. just curious if this functionality is something that we can leverage elsewhere, but it's a little hard to follow what's happening.

Reply:
Well, the steps of NumEncoder was shown in the jupyter notebook. For much clear showing, I'll add some necessary comments in the code.